cohort-engine/src/main/java/com/ibm/cohort/engine/measure/cache/CacheKey.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (C) Copyright IBM Corp. 2020, 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.cohort.engine.measure.cache;
+
+import com.ibm.cohort.annotations.Generated;
+import org.opencds.cqf.cql.engine.runtime.Code;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A HashMap friendly model representing the cacheable fields of `RetrieveProvider.retrieve()`.
+ * @see org.opencds.cqf.cql.engine.retrieve.RetrieveProvider
+ */
+public class CacheKey {
+
+	public static CacheKey create(
+			String context,
+			String contextPath,
+			String contextValue,
+			String dataType,
+			String templateId,
+			String codePath,
+			Iterable<Code> codes,
+			String valueSet
+	) {
+		List<CacheCode> cacheCodes = Collections.emptyList();
+		if (codes != null) {
+			cacheCodes = new ArrayList<>();
+			for (Code code : codes) {
+				cacheCodes.add(CacheCode.create(code));
+			}
+		}
+		return new CacheKey(context, contextPath, contextValue, dataType, templateId, codePath, cacheCodes, valueSet);
+	}
+
+	private final String context;
+	private final String contextPath;
+	private final String contextValue;
+	private final String dataType;
+	private final String templateId;
+	private final String codePath;
+	private final List<CacheCode> codes;
+	private final String valueSet;
+
+	public CacheKey(
+			String context,
+			String contextPath,
+			String contextValue,
+			String dataType,
+			String templateId,
+			String codePath,
+			List<CacheCode> codes,
+			String valueSet
+	) {
+		this.context = context;
+		this.contextPath = contextPath;
+		this.contextValue = contextValue;
+		this.dataType = dataType;
+		this.templateId = templateId;
+		this.codePath = codePath;
+		this.codes = codes;
+		this.valueSet = valueSet;
+	}
+
+	@Override
+	@Generated
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) return false;
+		CacheKey cacheKey = (CacheKey) o;
+		return Objects.equals(context, cacheKey.context)
+				&& Objects.equals(contextPath, cacheKey.contextPath)
+				&& Objects.equals(contextValue, cacheKey.contextValue)
+				&& Objects.equals(dataType, cacheKey.dataType)
+				&& Objects.equals(templateId, cacheKey.templateId)
+				&& Objects.equals(codePath, cacheKey.codePath)
+				&& Objects.equals(codes, cacheKey.codes)
+				&& Objects.equals(valueSet, cacheKey.valueSet);
+	}
+
+	@Override
+	@Generated
+	public int hashCode() {
+		return Objects.hash(context, contextPath, contextValue, dataType, templateId, codePath, codes, valueSet);
```

Comment:
When two lists of codes have the  same values in different  orders, do you want them to hash to the same value?

Reply:
Good question.  Since I already have to convert their `Iterable` to some concrete collection type.  Using a `Set` instead of a `List` will allow for codes in any order.  I just have to think through if there's any other implications to that.